### ISSUE TITLE
Does not compile with OpenSSL 1.1.0

### ISSUE THREAD

**Issue by sesse** [grpc/grpc#8801]
gRPC uses interfaces that are removed and/or deprecated in OpenSSL 1.1.0, so it does not compile (in particular, the RSA struct is now opaque). Please see the patch below. I have signed the CLA, but haven't been able to actually test the resulting code because of nanopb compilation issues.

```
Index: grpc-1.0.0/src/core/lib/security/credentials/jwt/jwt_verifier.c
===================================================================
--- grpc-1.0.0.orig/src/core/lib/security/credentials/jwt/jwt_verifier.c
+++ grpc-1.0.0/src/core/lib/security/credentials/jwt/jwt_verifier.c
@@ -467,6 +467,7 @@ static EVP_PKEY *pkey_from_jwk(const grp
   const grpc_json *key_prop;
   RSA *rsa = NULL;
   EVP_PKEY *result = NULL;
+  BIGNUM *n = NULL, *e = NULL;

   GPR_ASSERT(kty != NULL && json != NULL);
   if (strcmp(kty, "RSA") != 0) {
@@ -480,22 +481,30 @@ static EVP_PKEY *pkey_from_jwk(const grp
   }
   for (key_prop = json->child; key_prop != NULL; key_prop = key_prop->next) {
     if (strcmp(key_prop->key, "n") == 0) {
-      rsa->n = bignum_from_base64(validate_string_field(key_prop, "n"));
-      if (rsa->n == NULL) goto end;
+      n = bignum_from_base64(validate_string_field(key_prop, "n"));
+      if (n == NULL) goto end;
     } else if (strcmp(key_prop->key, "e") == 0) {
-      rsa->e = bignum_from_base64(validate_string_field(key_prop, "e"));
-      if (rsa->e == NULL) goto end;
+      e = bignum_from_base64(validate_string_field(key_prop, "e"));
+      if (e == NULL) goto end;
     }
   }
-  if (rsa->e == NULL || rsa->n == NULL) {
+  if (e == NULL || n == NULL) {
     gpr_log(GPR_ERROR, "Missing RSA public key field.");
     goto end;
   }
+  if (RSA_set0_key(rsa, n, e, NULL)) {
+    n = e = NULL;  // Now owned by the RSA object.
+  } else {
+    goto end;
+  }
+
   result = EVP_PKEY_new();
   EVP_PKEY_set1_RSA(result, rsa); /* uprefs rsa. */

 end:
   if (rsa != NULL) RSA_free(rsa);
+  if (n != NULL) BN_free(n);
+  if (e != NULL) BN_free(e);
   return result;
 }

Index: grpc-1.0.0/src/core/lib/tsi/ssl_transport_security.c
===================================================================
--- grpc-1.0.0.orig/src/core/lib/tsi/ssl_transport_security.c
+++ grpc-1.0.0/src/core/lib/tsi/ssl_transport_security.c
@@ -1330,7 +1330,7 @@ tsi_result tsi_create_ssl_client_handsha
   *factory = NULL;
   if (pem_root_certs == NULL) return TSI_INVALID_ARGUMENT;

-  ssl_context = SSL_CTX_new(TLSv1_2_method());
+  ssl_context = SSL_CTX_new(TLS_method());
   if (ssl_context == NULL) {
     gpr_log(GPR_ERROR, "Could not create ssl context.");
     return TSI_INVALID_ARGUMENT;
@@ -1341,6 +1341,12 @@ tsi_result tsi_create_ssl_client_handsha
   impl->ssl_context = ssl_context;

   do {
+    result = SSL_CTX_set_min_proto_version(ssl_context, TLS1_2_VERSION);
+    if (result != TSI_OK) {
+      gpr_log(GPR_ERROR, "Could not set minimum TLS version.");
+      break;
+    }
+
     result =
         populate_ssl_context(ssl_context, pem_private_key, pem_private_key_size,
                              pem_cert_chain, pem_cert_chain_size, cipher_list);
@@ -1461,12 +1467,17 @@ tsi_result tsi_create_ssl_server_handsha

   for (i = 0; i < key_cert_pair_count; i++) {
     do {
-      impl->ssl_contexts[i] = SSL_CTX_new(TLSv1_2_method());
+      impl->ssl_contexts[i] = SSL_CTX_new(TLS_method());
       if (impl->ssl_contexts[i] == NULL) {
         gpr_log(GPR_ERROR, "Could not create ssl context.");
         result = TSI_OUT_OF_RESOURCES;
         break;
       }
+      result = SSL_CTX_set_min_proto_version(impl->ssl_contexts[i], TLS1_2_VERSION);
+      if (result != TSI_OK) {
+        gpr_log(GPR_ERROR, "Could not set minimum TLS version.");
+        break;
+      }
       result = populate_ssl_context(
           impl->ssl_contexts[i], pem_private_keys[i], pem_private_keys_sizes[i],
           pem_cert_chains[i], pem_cert_chains_sizes[i], cipher_list);
```

**Comment by danieljoos** [comment]
It would be awesome to get this into grpc. I have a C++ project that already requires openssl 1.1 for some of its other dependencies.

**Comment by randomguy3** [comment]
I'm almost there with my branch - just some test failures to track down.

**Comment by pravi** [comment]
Any update on this? I'm packaging ruby grpc for debian and debian has openssl 1.1 as default.

**Comment by sesse** [comment]
Debian's gRPC is already patched, though. (I updated gRPC in Debian through an NMU; seemingly the package is effectively orphaned.)

**Comment by pravi** [comment]
@sesse but building ruby-grpc fails when libssl-dev is used (I have to use libssl1.0-dev). See my current work at https://anonscm.debian.org/cgit/pkg-ruby-extras/ruby-grpc.git (it also needs change in gem2deb because it has both native and pure ruby libs and gem2deb install only the native lib).

**Comment by dionescu** [comment]
Any update here? I'm running the latest release (1.4.2) and Bazel build is failing due to this issue.

**Comment by jtattermusch** [comment]
CC @jboeuf

**Comment by jtattermusch** [comment]
Can someone create a PR with a fix (the issue log looks like there's been several fixes or partial fixes toward grpc forks)? We need to maintain the ability to compile with both older and newer versions of OpenSSL though.

**Comment by jboeuf** [comment]
@justinburke FYI

I believe that https://github.com/grpc/grpc/pull/11496 fixes the issue but it needs to be updated. @agc-sec can you please (re)update your branch and @nicolasnoble can you please approve the changes once it's done so that Allen can merge? Thanks!

**Comment by elivings** [comment]
Any updates on this?

**Comment by a11r** [comment]
Fixed by #11495
